src/condor_negotiator.V6/matchmaker.cpp

```diff
@@ -1209,6 +1215,185 @@ compute_significant_attrs(ClassAdListDoesNotDeleteAds & startdAds)
+    //   - remote user is over quota AND request and remote users have ancestor
+    //     brothers, AND the request ABG is under quota, AND the remote ABG is
+    //     over quota
+
+    // enabled?
+    dprintf(D_FULLDEBUG, "Checking quota of request %s, remote %s\n",
+            request_user, remote_user);
+    if (!preemption_quota) {
+        dprintf(D_FULLDEBUG, "Quota Preemption disabled\n");
+        return false;
+    }
+
+    GroupEntry* request_group = accountant.GetAssignedGroup(request_user);
+    GroupEntry* remote_group = accountant.GetAssignedGroup(remote_user);
+
+    // root group special case
+    if (request_group->name == hgq_root_name) {
+        // root group cannot quota-preempt
+		dprintf(D_FULLDEBUG, "request group is root, cannot preempt\n");
+        return false;
+    }
+
+    // is remote user's group is not over quota, we are done
+    if (!isOverQuota(remote_group)) {
+        dprintf(D_FULLDEBUG, "remote group %s is not over\n", remote_group->name.c_str());
+        return false;
+    }
+
+    bool request_under_quota = isUnderQuota(request_group);
+
+    GroupEntry *request_abg, *remote_abg;
+    if (!getAncestorBrotherGroups(request_group, remote_group, request_abg,
+                                  remote_abg)) 
+    {
+        // same branch: only if request is under quota
+        dprintf(D_FULLDEBUG, "same branch and request group under is %d\n", 
+                request_under_quota);
+        return request_under_quota;
+    } else {
+        // different branches: decide according to ancestor brother groups
+        dprintf(D_FULLDEBUG, "different branches\n");
+        return isUnderQuota(request_abg) && isOverQuota(remote_abg);
+    }
+}
+
+bool Matchmaker::getAncestorBrotherGroups(GroupEntry* g1, GroupEntry* g2,
+                                          GroupEntry*& abg1, GroupEntry*& abg2)
+{
+	std::deque<GroupEntry*> g1_branch, g2_branch;
+    for (GroupEntry* ancestor = g1; ancestor != NULL;
+         ancestor = ancestor->parent)
+    {
+        g1_branch.push_front(ancestor);
+    }
+    for (GroupEntry* ancestor = g2; ancestor != NULL;
+         ancestor = ancestor->parent)
+    {
+        g2_branch.push_front(ancestor);
+    }
+
+    for (size_t i = 0; i < std::min(g1_branch.size(), g2_branch.size()); ++i) {
+        if (g1_branch[i] != g2_branch[i]) {
+            abg1 = g1_branch[i];
+            abg2 = g2_branch[i];
+            dprintf(D_FULLDEBUG, "ancestors: g1 %s, g2 %s, abg1 %s, abg2 %s\n",
+                    g1->name.c_str(), g2->name.c_str(), abg1->name.c_str(), 
+		    abg2->name.c_str());
+            return true;
+        }
+    }
+    dprintf(D_FULLDEBUG, "ancestors: g1 %s, g2 %s, same branch\n", 
+	    g1->name.c_str(), g2->name.c_str());
+    return false;
+}
+
+bool Matchmaker::isOverQuota(GroupEntry* group)
+{
+    float quota = group->quota;
+    float usage = 0.0;
+    if (!groupUsageHash->lookup(group->name, usage) == -1) {
```

Comment:
Compilation warning:
warning: logical not is only applied to the left hand side of comparison [-Wlogical-not-parentheses]
note: add parentheses around left hand side expression to silence this warning